server/etcdserver/api/v3discovery/discovery.go

```diff
@@ -0,0 +1,544 @@
+	}
+
+	return members, resp.Header.Revision, nil
+}
+
+func (d *discovery) checkClusterRetry() (memberInfoMap, int, int64, error) {
+	if d.retries < nRetries {
+		d.logAndBackoffForRetry("cluster status check")
+		return d.checkCluster()
+	}
+	return nil, 0, 0, ErrTooManyRetries
+}
+
+func (d *discovery) checkCluster() (memberInfoMap, int, int64, error) {
+	clusterSize, err := d.getClusterSize()
+	if err != nil {
+		if err == ErrSizeNotFound || err == ErrBadSizeKey {
+			return nil, 0, 0, err
+		}
+
+		return d.checkClusterRetry()
+	}
+
+	members, rev, err := d.getClusterMembers()
+	if err != nil {
+		return d.checkClusterRetry()
+	}
+
+	// find self position
+	memberSelfId := getMemberKey(d.cluster, d.id.String())
+	idx := 0
+	for k := range members {
+		if k == memberSelfId {
+			break
+		}
+		if idx >= clusterSize-1 {
+			return members, clusterSize, rev, ErrFullCluster
+		}
+		idx++
+	}
+	return members, clusterSize, rev, nil
+}
+
+func (d *discovery) registerSelfRetry(contents string) error {
+	if d.retries < nRetries {
+		d.logAndBackoffForRetry("register member itself")
+		return d.registerSelf(contents)
+	}
+	return ErrTooManyRetries
+}
+
+func (d *discovery) registerSelf(contents string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), d.cfg.RequestTimeOut)
+	memberKey := getMemberKey(d.cluster, d.id.String())
+	_, err := d.c.Put(ctx, memberKey, contents)
+	cancel()
+
+	if err != nil {
+		d.lg.Warn(
+			"failed to register members itself to the discovery service",
+			zap.String("discovery-url", d.durl),
+			zap.String("memberKey", memberKey),
+			zap.Error(err),
+		)
+		return d.registerSelfRetry(contents)
+	}
+
+	d.lg.Info(
+		"register member itself successfully",
+		zap.String("discovery-url", d.durl),
+		zap.String("memberKey", memberKey),
+		zap.String("memberInfo", contents),
+	)
+
+	return nil
+}
+
+func (d *discovery) waitNodes(members memberInfoMap, clusterSize int, rev int64) (memberInfoMap, error) {
+	// watch from the next revision
+	membersKeyPrefix := getMemberKeyPrefix(d.cluster)
+	w := d.c.Watch(context.Background(), membersKeyPrefix, clientv3.WithPrefix(), clientv3.WithRev(rev+1))
+
+	d.lg.Info(
+		"waiting for peers from discovery service",
+		zap.String("discovery-url", d.durl),
+		zap.Int("clusterSize", clusterSize),
+		zap.Int("found-peers", len(members)),
+	)
+
+	// waiting for peers until all needed peers are returned
+	for wresp := range w {
+		for _, ev := range wresp.Events {
+			mKey := strings.TrimSpace(string(ev.Kv.Key))
+			mValue := strings.TrimSpace(string(ev.Kv.Value))
+
+			if shouldIgnorePeer(members, mKey, mValue, d.cluster, d.lg, d.durl) {
+				continue
+			}
+
+			d.lg.Info(
+				"found peer from discovery service",
+				zap.String("discovery-url", d.durl),
+				zap.String("memberKey", mKey),
+				zap.String("memberInfo", mValue),
+			)
+
+			members[mKey] = memberInfo{peerURL: mValue, createRev: ev.Kv.CreateRevision}
+		}
+
+		if len(members) >= clusterSize {
```

Comment:
This can result with number of members larger then cluster size, is this intentional?

Reply:
Yes, it's intentional.  Multiple members may try to join the cluster concurrently, so the number of members may be larger than the cluster size.  But we sort all the member based on createRevision, and only get the first ${clusterSize} members. 

If a members registered itself successfully, but there are always ${clusterSize} registered members (all with smaller CreateRevisions then the member), then the member will get a ErrFullCluster error. 

Please let me know if you need further clarification.